pkg/registry/core/pod/strategy.go

```diff
@@ -287,44 +305,45 @@ var ResizeStrategy = podResizeStrategy{
+		fieldpath.MakePrefixMatcherOrDie("spec", "initContainers", fieldpath.MatchAnyPathElement(), "resources"),
+		fieldpath.MakePrefixMatcherOrDie("spec", "initContainers", fieldpath.MatchAnyPathElement(), "resizePolicy"),
 	),
 }
 
 // dropNonResizeUpdates discards all changes except for pod.Spec.Containers[*].Resources, pod.Spec.InitContainers[*].Resources, ResizePolicy and certain metadata
 func dropNonResizeUpdates(newPod, oldPod *api.Pod) *api.Pod {
-	pod := dropPodUpdates(newPod, oldPod)
-
-	// Containers are not allowed to be re-ordered, but in case they were,
-	// we don't want to corrupt them here. It will get caught in validation.
-	oldCtrToIndex := make(map[string]int)
-	oldInitCtrToIndex := make(map[string]int)
-	for idx, ctr := range pod.Spec.Containers {
-		oldCtrToIndex[ctr.Name] = idx
-	}
-	for idx, ctr := range pod.Spec.InitContainers {
-		oldInitCtrToIndex[ctr.Name] = idx
+	// Containers are not allowed to be added, removed, re-ordered, or renamed.
+	// If we detect any of these changes, we will return new podspec as-is and
+	// allow the validation to catch the error and drop the update.
+	if len(newPod.Spec.Containers) != len(oldPod.Spec.Containers) || len(newPod.Spec.InitContainers) != len(oldPod.Spec.InitContainers) {
+		return newPod
 	}
 
-	for _, ctr := range newPod.Spec.Containers {
-		idx, ok := oldCtrToIndex[ctr.Name]
-		if !ok {
-			continue
-		}
-		pod.Spec.Containers[idx].Resources = ctr.Resources
-		pod.Spec.Containers[idx].ResizePolicy = ctr.ResizePolicy
+	oldPodCopyWithMergedResources := oldPod.DeepCopy()
+	oldPodCopyWithMergedResources.Spec.Containers = dropNonResizeUpdatesForContainers(newPod.Spec.Containers, oldPodCopyWithMergedResources.Spec.Containers)
+	if utilfeature.DefaultFeatureGate.Enabled(features.SidecarContainers) {
+		oldPodCopyWithMergedResources.Spec.InitContainers = dropNonResizeUpdatesForContainers(newPod.Spec.InitContainers, oldPodCopyWithMergedResources.Spec.InitContainers)
 	}
 
-	if utilfeature.DefaultFeatureGate.Enabled(features.SidecarContainers) {
-		for _, ctr := range newPod.Spec.InitContainers {
-			idx, ok := oldInitCtrToIndex[ctr.Name]
-			if !ok {
-				continue
-			}
-			pod.Spec.InitContainers[idx].Resources = ctr.Resources
-			pod.Spec.InitContainers[idx].ResizePolicy = ctr.ResizePolicy
+	newPod.Spec = oldPod.Spec
+	newPod.Status = oldPod.Status
+	metav1.ResetObjectMetaForStatus(&newPod.ObjectMeta, &oldPod.ObjectMeta)
+	newPod.Spec.Containers = oldPodCopyWithMergedResources.Spec.Containers
+	newPod.Spec.InitContainers = oldPodCopyWithMergedResources.Spec.InitContainers
+	return newPod
+}
+
+func dropNonResizeUpdatesForContainers(new, old []api.Container) []api.Container {
+	for i, ctr := range new {
+		if old[i].Name != new[i].Name {
+			// This is an attempt to reorder or rename a container, which is not allowed.
+			// Allow validation to catch this error.
+			return new
 		}
+		old[i].Resources = ctr.Resources
+		old[i].ResizePolicy = ctr.ResizePolicy
```

Comment:
https://github.com/kubernetes/kubernetes/pull/130183 made it so that ResizePolicy is no longer mutable. Should we drop the update now, or continue to fail it in validation?

Reply:
Personally I think it would be safer to continue to fail it in validation. Granted I may be missing some context, but I could see a user wanting to bundle together an update to container resources and a change to the resizePolicy, and they may find it surprising to see it only partially go through? 

If we start with this today, you can also always relax this validation failure later.